comment_iowriter.go

```diff
@@ -142,3 +144,135 @@ func (cw *RDJSONCommentWriter) Flush(_ context.Context) error {
 	}
 	return nil
 }
+
+var _ CommentService = &SARIFCommentWriter{}
+
+// SARIFCommentWriter
+type SARIFCommentWriter struct {
+	w        io.Writer
+	comments []*Comment
+	toolName string
+}
+
+func NewSARIFCommentWriter(w io.Writer, toolName string) *SARIFCommentWriter {
+	return &SARIFCommentWriter{w: w, toolName: toolName}
+}
+
+func (cw *SARIFCommentWriter) Post(_ context.Context, c *Comment) error {
+	cw.comments = append(cw.comments, c)
+	return nil
+}
+
+func (cw *SARIFCommentWriter) Flush(_ context.Context) error {
+	run := sarif.Run{
+		Tool: sarif.Tool{
+			Driver: sarif.ToolComponent{
+				Name: cw.toolName,
+			},
+		},
+	}
+	rules := make(map[string]sarif.ReportingDescriptor)
+	for _, c := range cw.comments {
+		result := sarif.Result{
+			Message: sarif.Message{
+				Text: sarif.String(c.Result.Diagnostic.Message),
+			},
+		}
+		if code := c.Result.Diagnostic.GetCode(); code.GetValue() != "" {
+			result.RuleID = sarif.String(code.GetValue())
+			rules[code.GetValue()] = sarif.ReportingDescriptor{
+				ID:      code.GetValue(),
+				HelpURI: sarif.String(code.GetUrl()),
+			}
```

Comment:
```suggestion
			run.Tool.Driver.Rules = append(run.Tool.Driver.Rules, sarif.ReportingDescriptor{
				ID:      code.GetValue(),
				HelpURI: sarif.String(code.GetUrl()),
			})
```

Values of the variable `rules` are only used when assigning to `run.Tool.Driver.Rules`.
Therefore, I think it is better to assign to `run.Tool.Driver.Rules` here.

Reply:
There can be multiple results with the same Rule ID. 
We should add one rule per rule id.

Reply:
On second thought, yes, we can refactore the code a bit to make it clear. 
I refactore the code. 907b2c0d5e4fcc98aed49d09ca3426b306e1dd18 
Thanks for the suggestion!